Fix regression introduced by #9960

shindere opened the pull request:
This commit restores the compiler's behaviour when called on a C file
and with no "-o" otption.

Octachron commented on Changes:
```diff
@@ -486,9 +486,9 @@ OCaml 4.13.0
-  (Sébastien Hinderer, reported by Daniel Bünzli, review by Florian
-  Angeletti and Gabriel Scherer)
+- #9960, #10619: extend ocamlc/ocamlopt's -o option to work when
+  compiling C files (Sébastien Hinderer, reported by Daniel Bünzli,
```
The authors and reviewer information should be on its own separate line.

shindere commented:
Florian Angeletti (2021/09/08 07:09 -0700):
> @Octachron commented on this pull request.
>
>
>
> > @@ -486,9 +486,9 @@ OCaml 4.13.0
>  - #9407: added warning for missing mli interface file
>    (Anukriti Kumar, review by Florian Angeletti)
>
> -- #9960: extend ocamlc/ocamlopt's -o option to work when compiling C files
> -  (Sébastien Hinderer, reported by Daniel Bünzli, review by Florian
> -  Angeletti and Gabriel Scherer)
> +- #9960, #10619: extend ocamlc/ocamlopt's -o option to work when
> +  compiling C files (Sébastien Hinderer, reported by Daniel Bünzli,
>
> The authors and reviewer information should be on its own separate
> line.

Does it look better to you like this?

Octachron commented on driver/compenv.ml:
```diff
@@ -641,13 +641,18 @@ let process_action
-        | None -> c_object_of_filename name
-        | Some n -> n
-      in
-      if Ccomp.compile_file ~output:obj_name name <> 0
```
An alternative implementation would be to keep the definition of `obj_name` and replace the line above with
```ocaml
Ccomp.compile_file ?output:!output_name
```

Octachron reviewed (approved):
Outside of my minor implementation remark, this PR looks good: it is a conservative version of the existing code which makes sure to only use `-o ` flags that were supplied by the user.

shindere commented:
Florian Angeletti (2021/09/08 14:35 +0000):
> @Octachron commented on this pull request.
>
>
>
> > @@ -641,13 +641,18 @@ let process_action
>    | ProcessCFile name ->
>        readenv ppf (Before_compile name);
>        Location.input_name := name;
> -      let obj_name = match !output_name with
> -        | None -> c_object_of_filename name
> -        | Some n -> n
> -      in
> -      if Ccomp.compile_file ~output:obj_name name <> 0
>
> An alternative implementation would be to keep the definition of `obj_name` and replace the line above with
> ```ocaml
> Ccomp.compile_file ?output:!output_name
> ```

That's much better, you are right! I'm not so used to lables. Updating.

shindere commented:
Florian Angeletti (2021/09/08 14:39 +0000):
> @Octachron approved this pull request.
>
> Outside of my minor implementation remark, this PR looks good: it is a
> conservative version of the existing code which makes sure to only use
> `-o ` flags that were supplied by the user.

So, shall we merge as-is and keep the improvement and tests for later?

That would be my prefered solution, I have to say.

>

Octachron commented:
As you wish, but if you think that my proposition lead to more readable code, it might be better to amend the PR now rather than waiting for a subsequent improvement PR. Contrarily, I agree that the test can wait because they don't seem trivial to implement.

shindere committed:
Fix regression introduced by #9660

This commit restores the compiler's behaviour when called on a C file
and with no "-o" otption.

shindere commented:
Florian Angeletti (2021/09/08 07:59 -0700):
> As you wish, but if you think that my proposition lead to more
> readable code, it might be better to amend the PR now rather than
> waiting for a subsequent improvement PR. Contrarily, I agree that the
> test can wait because they don't seem trivial to implement.

I updated the code. It still passes the testsuite.

shindere commented:
Thanks a lot for the merge and the support!
